File: plugins/openclaw/rana-plugin.js
/**
 * rana-plugin.js — RanA marker bridge for OpenClaw
 * ------------------------------------------------------------------
 * Emits run-lifecycle MARKERS to the local RanA marker socket so the RanA
 * timeline can render "inbound message -> these execs -> these egress connects
 * -> this file changed" (causality), instead of a raw undifferentiated event
 * river.
 *
 * WHAT THIS SENDS (and the hard rule it enforces):
 *   Identifiers + lifecycle ONLY: { runId, agentId, channel, status, phase, ts }
 *   NEVER message text, prompts, completions, summaries, or any content.
 * This mirrors RanA principle P7 ("effects, not thoughts") and P1 (markers are
 * ENRICHMENT — origin=enrichment; a marker can mislabel, never fabricate a
 * kernel event). The redaction below is defense-in-depth: even if a future
 * OpenClaw event shape leaks content into a field we forward, the allowlist
 * strips it before it leaves this process.
 *
 * TRANSPORT: newline-delimited JSON over a Unix domain socket whose path +
 * per-session token are provided by RanA at `rana adopt openclaw` time via the
 * RANA_MARKER_SOCKET / RANA_MARKER_TOKEN environment variables. If they are
 * absent, the plugin is inert (no socket, no error) — RanA simply falls back to
 * inferred causality.
 *
 * INSTALL: drop into OpenClaw's plugin directory (see docs/OPENCLAW.md). RanA's
 * `rana adopt openclaw` offers to do this for you and to set the env vars on the
 * gateway unit. Consent is prompted; default yes; fully optional.
 *
 * LICENSE: Apache-2.0 (same as RanA).
 */

'use strict';

const net = require('node:net');

// ---- The allowlist. Only these keys may ever leave this process. -----------
const ALLOWED_FIELDS = ['runId', 'agentId', 'channel', 'status', 'phase', 'ts'];
// Belt and suspenders: keys we explicitly refuse even if someone adds them.
const FORBIDDEN_FIELDS = ['text', 'prompt', 'completion', 'message', 'content', 'summary', 'body'];

const SOCKET_PATH = process.env.RANA_MARKER_SOCKET || null;
const SOCKET_TOKEN = process.env.RANA_MARKER_TOKEN || '';

/** Reduce an arbitrary object to the allowlisted, content-free marker payload. */
function sanitize(obj) {
  const out = {};
  for (const k of ALLOWED_FIELDS) {
    if (obj[k] === undefined || obj[k] === null) continue;
    if (FORBIDDEN_FIELDS.includes(k)) continue; // unreachable given the allowlist, kept as intent
    const v = obj[k];
    // Only primitives survive; no nested objects that could smuggle content.
    if (typeof v === 'string' || typeof v === 'number' || typeof v === 'boolean') {
      out[k] = v;
    }
  }
  return out;
}

/**
 * Fire-and-forget marker send. Never throws into the gateway, never blocks the
 * agent loop (P2: enrichment must not affect the workload). One short-lived
 * connection per marker keeps the plugin stateless and crash-proof.
 */
function sendMarker(kind, fields) {
  if (!SOCKET_PATH) return; // inert when RanA isn't present
  let payload;
  try {
    payload = JSON.stringify({
      v: 1,
      token: SOCKET_TOKEN,
      event: kind,                // "run.start" | "run.end" — the listener keys
                                  // the marker.<event> type off the "event"
                                  // field (internal/service/marker.go), NOT "kind".
      origin: 'enrichment',       // P1: never authoritative
      ...sanitize(fields),
    }) + '\n';
  } catch {
    return; // if we can't even serialize identifiers, drop silently
  }
  try {
    const sock = net.createConnection(SOCKET_PATH);
    sock.on('error', () => sock.destroy());   // RanA down? no-op.
    sock.on('connect', () => {
      sock.write(payload, () => sock.end());
    });
    // Hard timeout so a wedged socket never lingers.
    sock.setTimeout(250, () => sock.destroy());
  } catch {
    /* swallow — markers are best-effort by design */
  }
}

/**
 * OpenClaw plugin entry point.
 *
 * OpenClaw's agent run lifecycle (see docs.openclaw.ai/concepts/architecture):
 *   req:agent  -> res:agent ack   { runId, status: "accepted" }   => run.start
 *   event:agent (streaming...)                                     (ignored: content)
 *   res:agent final               { runId, status, summary }      => run.end
 *
 * We hook the accepted/final transitions and forward the runId + status. The
 * streaming content frames are deliberately NOT hooked.
 *
 * The exact registration API name may vary by OpenClaw version; adapt the two
 * `on(...)` bindings to the installed gateway's plugin surface. Everything
 * below the bindings is version-independent.
 */
module.exports = function register(api) {
  // Run accepted -> mark the start of a causal cluster.
  api.on('agent:accepted', (ev) => {
    sendMarker('run.start', {
      runId: ev.runId,
      agentId: ev.agentId,
      channel: ev.channel,
      status: 'accepted',
      phase: 'start',
      ts: Date.now(),
    });
  });

  // Run final -> close the cluster. Note we forward `status` (ok/error) but
  // NOT `summary`, which may contain model-authored content.
  api.on('agent:final', (ev) => {
    sendMarker('run.end', {
      runId: ev.runId,
      agentId: ev.agentId,
      channel: ev.channel,
      status: ev.status,   // e.g. "ok" | "error" — a label, not content
      phase: 'end',
      ts: Date.now(),
    });
  });

  return {
    name: 'rana-marker-bridge',
    version: '1.0.0',
    description: 'Emits content-free run-lifecycle markers to RanA for causality.',
  };
};

File: plugins/hermes/rana_plugin.py
"""
RanA marker emitter for Hermes Agent (Nous Research).

Emits run-lifecycle MARKERS to the local RanA marker socket so the RanA
timeline can render "inbound message -> these execs -> these egress connects
-> this file changed" — clustering a run's kernel effects by runId.

WHAT IT SENDS (and NOTHING else): identifiers + lifecycle ONLY —
{ runId, agentId, channel, status, phase, ts }. It NEVER sends prompts,
completions, message text, or any model I/O (P7, RanA's #1 scope wall). The
sanitizer below drops everything outside the allowlist before it leaves this
process, and RanA's own marker listener rejects any line carrying a forbidden
field, so a bug here fails safe.

TRANSPORT: newline-delimited JSON over a Unix domain socket whose path + a
per-session token are provided by RanA via the RANA_MARKER_SOCKET /
RANA_MARKER_TOKEN environment variables. If RANA_MARKER_SOCKET is absent, the
plugin is completely INERT (no socket, no error) — RanA simply isn't present.

INSTALL: drop this file in ~/.hermes/plugins/ (see README.md). Hermes discovers
plugins there and calls register(context); this plugin registers hooks on the
run/session start and end lifecycle events. Because the exact hook names differ
across Hermes versions, register() is defensive and best-effort — the
protocol-critical core (send_marker / sanitize) is version-independent.
"""

import json
import os
import socket
import time

# The ONLY keys allowed to leave this process. Anything else is dropped.
ALLOWED_FIELDS = ("runId", "agentId", "channel", "status", "phase", "ts")

SOCKET_PATH = os.environ.get("RANA_MARKER_SOCKET") or None
SOCKET_TOKEN = os.environ.get("RANA_MARKER_TOKEN", "")

_CONNECT_TIMEOUT = 0.25  # seconds; a wedged socket must never linger


def _sanitize(fields):
    """Keep only allowlisted identifier/lifecycle fields; coerce to scalars."""
    out = {}
    for k in ALLOWED_FIELDS:
        if k in fields and fields[k] is not None:
            v = fields[k]
            if isinstance(v, (str, int, float, bool)):
                out[k] = v
            else:
                out[k] = str(v)
    return out


def send_marker(event, **fields):
    """
    Emit one marker. event is "run.start" or "run.end". Best-effort and
    crash-proof: any failure (RanA down, serialize error, wedged socket) is
    swallowed silently. One connection per marker keeps this stateless.
    """
    if not SOCKET_PATH:
        return  # inert when RanA isn't present
    try:
        payload = json.dumps(
            {
                "v": 1,
                "token": SOCKET_TOKEN,
                # The listener keys the marker.<event> type off the "event"
                # field (internal/service/marker.go), NOT "kind".
                "event": event,
                "origin": "enrichment",  # P1: markers are never authoritative
                **_sanitize(fields),
            },
            separators=(",", ":"),
        ) + "\n"
    except Exception:
        return  # if we can't even serialize identifiers, drop silently

    sock = None
    try:
        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        sock.settimeout(_CONNECT_TIMEOUT)
        sock.connect(SOCKET_PATH)
        sock.sendall(payload.encode("utf-8"))
    except Exception:
        pass  # swallow — markers are best-effort by design
    finally:
        if sock is not None:
            try:
                sock.close()
            except Exception:
                pass


def _run_id(ctx):
    """Best-effort extraction of a stable per-run id from a Hermes event/ctx."""
    for attr in ("run_id", "runId", "session_id", "sessionId", "conversation_id"):
        v = getattr(ctx, attr, None) or (ctx.get(attr) if isinstance(ctx, dict) else None)
        if v:
            return str(v)
    return None


def _channel(ctx):
    for attr in ("channel", "surface", "source"):
        v = getattr(ctx, attr, None) or (ctx.get(attr) if isinstance(ctx, dict) else None)
        if v:
            return str(v)
    return None


def on_run_start(ctx=None):
    """Hook: a run/conversation began."""
    send_marker(
        "run.start",
        runId=_run_id(ctx),
        agentId="hermes",
        channel=_channel(ctx),
        status="accepted",
        phase="start",
        ts=int(time.time() * 1000),
    )


def on_run_end(ctx=None, status="completed"):
    """Hook: a run/conversation finished."""
    send_marker(
        "run.end",
        runId=_run_id(ctx),
        agentId="hermes",
        status=status,
        phase="end",
        ts=int(time.time() * 1000),
    )


def register(context):
    """
    Hermes plugin entry point. Hermes discovers plugins under ~/.hermes/plugins/
    and calls register(context); the context exposes a hook-registration API.
    Because hook names vary across Hermes versions, this tries the common ones
    and no-ops on whatever isn't present (never raises — a plugin must not break
    the agent). Adjust the event names below to your Hermes version's
    gateway/hooks.py lifecycle if they differ (see README.md).
    """
    if context is None:
        return

    # Prefer an explicit hook-registration method if the context exposes one.
    register_hook = getattr(context, "register_hook", None) or getattr(context, "add_hook", None)
    if callable(register_hook):
        for name in ("run_start", "session_start", "conversation_start", "on_run_start"):
            try:
                register_hook(name, on_run_start)
            except Exception:
                pass
        for name in ("run_end", "session_end", "conversation_end", "on_run_end"):
            try:
                register_hook(name, on_run_end)
            except Exception:
                pass
        return

    # Fallback: a generic event bus (context.on(event, handler)).
    on = getattr(context, "on", None)
    if callable(on):
        try:
            on("run.start", on_run_start)
            on("run.end", lambda c=None: on_run_end(c))
        except Exception:
            pass


# Manual smoke test: `python rana_plugin.py` sends a run.start/run.end pair to
# $RANA_MARKER_SOCKET (start a listener there first) so you can verify the wire
# format against a loopback socket without a running Hermes.
if __name__ == "__main__":
    on_run_start({"run_id": "smoke-0001", "channel": "cli"})
    on_run_end({"run_id": "smoke-0001"}, status="completed")
    print("sent run.start + run.end to", SOCKET_PATH or "(inert: RANA_MARKER_SOCKET unset)")
